contracts/FlowStakingCollection.cdc

```diff
@@ -0,0 +1,844 @@
+                    staker.requestUnstaking(amount: amount)
+                }
+            }
+        }
+
+        /// Function to unstake all tokens for an existing node staking record in the StakingCollection
+        /// Only available for node operators
+        pub fun unstakeAll(nodeID: String) {
+            pre {
+                self.doesStakeExist(nodeID: nodeID, delegatorID: nil): "Specified stake does not exist"
+            }
+    
+            if let node = self.borrowNode(nodeID) {
+                node.unstakeAll()
+            } else {
+                let staker = self.tokenHolder!.borrow()!.borrowStaker()
+                staker.unstakeAll()
+            }
+        }
+
+        /// Function to withdraw unstaked tokens for an existing Stake or Delegation record in the StakingCollection
+        pub fun withdrawUnstakedTokens(nodeID: String, delegatorID: UInt32?, amount: UFix64) { 
+            pre {
+                self.doesStakeExist(nodeID: nodeID, delegatorID: delegatorID): "Specified stake does not exist"
+            }
+
+            if let _delegatorID = delegatorID {
+                if let delegator = self.borrowDelegator(nodeID, _delegatorID) {
+                    let tokens <- delegator.withdrawUnstakedTokens(amount: amount)
+                    self.depositTokens(from: <-tokens)
+                } else {
+                    let delegator = self.tokenHolder!.borrow()!.borrowDelegator()
+                    delegator.withdrawUnstakedTokens(amount: amount)
+                }
+            } else {
+                if let node = self.borrowNode(nodeID) {
+                    let tokens <- node.withdrawUnstakedTokens(amount: amount)
+                    self.depositTokens(from: <-tokens)
+                } else {
+                    let staker = self.tokenHolder!.borrow()!.borrowStaker()
+                    staker.withdrawUnstakedTokens(amount: amount)
+                }
+            }
+        }
+
+        /// Function to withdraw rewarded tokens for an existing Stake or Delegation record in the StakingCollection
+        pub fun withdrawRewardedTokens(nodeID: String, delegatorID: UInt32?, amount: UFix64) {
+            pre {
+                self.doesStakeExist(nodeID: nodeID, delegatorID: delegatorID): "Specified stake does not exist"
+            }
+
+            if let _delegatorID = delegatorID {
+                if let delegator = self.borrowDelegator(nodeID, _delegatorID) {
+                    // We update the unlocked tokens used field before withdrawing because 
+                    // rewards are newly minted and not immediately reflected in the tokens used fields
+                    self.unlockedTokensUsed = self.unlockedTokensUsed + amount
+
+                    let tokens <- delegator.withdrawRewardedTokens(amount: amount)
+
+                    self.depositTokens(from: <-tokens)
+                } else {
+                    let delegator = self.tokenHolder!.borrow()!.borrowDelegator()
```

Comment:
Maybe refactor the repeated part `self.tokenHolder!.borrow()!` out into a variable